src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyState = { ...state };
+  const newArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        for (const del in copyState) {
+          delete copyState[del];
+        }
+        newArray.push({ ...copyState });
+        break;
+
+      case 'addProperties':
+        Object.assign(copyState, action.extraData);
+        newArray.push({ ...copyState });
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete copyState[key];
+        }
+        newArray.push({ ...copyState });
+    }
+  }
```

Comment:
Don't repeat yourself. Move common commands outside separated scopes 
```suggestion
  for (const action of actions) {
    switch (action.type) {
      case 'clear':
        for (const del in copyState) {
          delete copyState[del];
        }
        break;

      case 'addProperties':
        Object.assign(copyState, action.extraData);
        break;

      case 'removeProperties':
        for (const key of action.keysToRemove) {
          delete copyState[key];
        }
    }
    
    newArray.push({ ...copyState });
  }
```

Reply:
I don't know yet what is forEch